qiskit/transpiler/passes/synthesis/high_level_synthesis.py

```diff
@@ -168,5 +193,160 @@ class DefaultSynthesisLinearFunction(HighLevelSynthesisPlugin):
+class PMHSynthesisLinearFunction(HighLevelSynthesisPlugin):
+    """Linear function synthesis plugin based on the Patel-Markov-Hayes method."""
+
+    def run(self, high_level_object, **options):
+        """Run synthesis for the given LinearFunction."""
+
+        # options supported by this plugin
+        coupling_map = options.get("coupling_map", None)
+        consider_all_mats = options.get("all_mats", 0)
+        consider_original_circuit = options.get("orig_circuit", 1)
+        optimize_count = options.get("opt_count", 1)
+
+        # At the end, if not none, represents the best decomposition.
+        best_decomposition = None
+
+        if consider_original_circuit:
+            best_decomposition = high_level_object.original_circuit
+
+        # This synthesis method is not aware of the coupling map, so we cannot apply
+        # this method when the coupling map is not None.
+        # (Though, technically, we could check if the reduced coupling map is
+        # fully-connected).
+
+        if not coupling_map:
+            if not consider_all_mats:
+                decomposition = synth_cnot_count_full_pmh(high_level_object.linear)
+            else:
+                decomposition = optimize_cx_4_options(
+                    synth_cnot_count_full_pmh,
+                    high_level_object.linear,
+                    optimize_count=optimize_count,
+                )
+
+            if not best_decomposition or _compare_circuits(
+                best_decomposition, decomposition, optimize_count=optimize_count
+            ):
+                best_decomposition = decomposition
+
+        return best_decomposition
+
+
+def _hamiltonian_paths(
+    coupling_map: CouplingMap, cutoff: Union[None | int] = None
+) -> List[List[int]]:
+    """Returns a list of all Hamiltonian paths in ``coupling_map`` (stopping the enumeration when
+    the number of already discovered paths exceeds the ``cutoff`` value, when specified).
+    In particular, returns an empty list if there are no Hamiltonian paths.
+    """
+
+    # This is a temporary function, the plan is to move it to rustworkx
+
+    def should_stop():
+        return cutoff is not None and len(all_paths) >= cutoff
+
+    def _recurse(current_node):
+        current_path.append(current_node)
+        if len(current_path) == coupling_map.size():
+            # Discovered a new Hamiltonian path
+            all_paths.append(current_path.copy())
+
+        if should_stop():
+            return
+
+        unvisited_neighbors = [
+            node for node in coupling_map.neighbors(current_node) if node not in current_path
+        ]
+        for node in unvisited_neighbors:
+            _recurse(node)
+            if should_stop():
+                return
+
+        current_path.pop()
+
+    all_paths = []
+    current_path = []
+    qubits = coupling_map.physical_qubits
+
+    for qubit in qubits:
+        _recurse(qubit)
+        if should_stop():
+            break
+    return all_paths
```

Comment:
It might be something that can implement with a specially crafted visitor, but I'm not sure. It's at least not obvious to me quickly scanning the docs and the code. Maybe @georgios-ts knows (as he wrote the functions), but yeah as a temporary step this is fine while waiting on a dedicated function in rustworkx.

Reply:
Thanks, @georgios-ts! I have already noticed the min-depth option, see my comment here: https://github.com/Qiskit/qiskit-terra/pull/9250#discussion_r1055433390. What scares me is what if there is an exponential number of hamiltonian paths, is there a way to limit rustworkx to compute only some fixed number of them?